[WIP] added basic machinery for defining coinductive types

cipher1024 opened the pull request.

cipher1024 committed:
rebased the inductive types on top of `wlog` PR

cipher1024 committed:
[WIP] broke down definitions and proofs into two namespaces: coinduction.approximation and cofix

cipher1024 committed:
[WIP] added example of usage of `data.coinductive`

cipher1024 committed:
[WIP] fixed syntax error

cipher1024 committed:
[WIP] added the construction of an infinite tree in the examples

cipher1024 committed:
[WIP]  in example, added a specialized corecursor

cipher1024 committed:
[WIP] fixed proofs in `data.coinductive`

gebner commented on data/coinductive.lean:
```diff
@@ -0,0 +1,712 @@
+-/
+import data.pfun
+
+import data.stream
```
This import is not needed, right?

gebner commented on data/coinductive.lean:
```diff
@@ -0,0 +1,712 @@
+  apply n_ih,
+end
+
+abbreviation path' (β : α → Type v) := list (Σ i, β i)
```
AFAIK, the `abbreviation` command is not intended for general usage.  Its only purpose is to define `id_rhs`.  The difference compared to `def` is that it sets a special reducibility hint for the kernel type-checker.

cipher1024 replied:
Thanks for the note! I thought I read a comment from Leo about `abbreviation` removing the need to ever use `reducible` but I can't find it anymore. In any case, this definition works even without `reducible` so I'll just make it plain.

gebner commented on data/coinductive.lean:
```diff
@@ -0,0 +1,712 @@
+variables {β}
+protected def corec (i : X) : cofix β :=
+{ approx := s_corec f i
+, consistent := P_corec _ _ }
```
I won't complain about coding style. https://github.com/leanprover/mathlib/blob/master/docs/style.md

cipher1024 replied:
I think the naming is good but the spacing probably needs work. The proofs I think need a lot of work. A lot are far too long. I'll be addressing both as I whip the package into shape

cipher1024 committed:
[WIP] small fixes and more examples

cipher1024 committed:
[WIP] In example, corecursive definitions are allowed to nest `l_two_tree` constructors arbitrarily deep

cipher1024 committed:
[WIP] added a `co_cases` tactic for manipulating the inner
representation of coinductive types.

Not meant for manipulating individual coinductive types

cipher1024 committed:
[WIP] remove the (now) redundant `hfunext`

cipher1024 committed:
[WIP] coinductive type families

cipher1024 committed:
[WIP] more families

cipher1024 committed:
[WIP] reason about `cofix` rather than `cofix_a`
